src/rust/src/sapling.rs

```diff
@@ -343,21 +366,51 @@ impl BatchValidator {
     /// `sighash` must be for the transaction this bundle is within.
     #[allow(clippy::boxed_local)]
     fn check_bundle(&mut self, bundle: Box<Bundle>, sighash: [u8; 32]) -> bool {
-        if let Some(validator) = &mut self.0 {
-            validator.check_bundle(bundle.0, sighash)
+        if let Some(inner) = &mut self.0 {
+            let cache = sapling_bundle_validity_cache();
+
+            // Compute the cache entry for this bundle.
+            let cache_entry = {
+                let bundle_commitment = bundle.commitment(TxIdDigester).unwrap();
+                let bundle_authorizing_commitment = bundle.commitment(BlockTxCommitmentDigester);
+                cache.compute_entry(
+                    bundle_commitment.as_bytes().try_into().unwrap(),
+                    bundle_authorizing_commitment.as_bytes().try_into().unwrap(),
+                    &sighash,
+                )
+            };
+
+            // Check if this bundle's validation result exists in the cache.
+            if cache.contains(cache_entry, &mut inner.queued_entries) {
+                true
+            } else {
+                // The bundle has been added to `inner.queued_entries` because it was not
+                // in the cache. We now check the bundle against the Sapling-specific
+                // consensus rules, and add its authorization to the validation batch.
+                inner.validator.check_bundle(bundle.0, sighash)
+            }
         } else {
             tracing::error!("sapling::BatchValidator has already been used");
             false
         }
     }
 
     fn validate(&mut self) -> bool {
-        if let Some(validator) = self.0.take() {
-            validator.validate(
+        if let Some(inner) = self.0.take() {
+            if inner.validator.validate(
                 unsafe { SAPLING_SPEND_VK.as_ref() }.unwrap(),
                 unsafe { SAPLING_OUTPUT_VK.as_ref() }.unwrap(),
                 OsRng,
-            )
+            ) {
+                // `Self::validate()` is only called if every `Self::check_bundle()`
+                // returned `true`, so at this point every bundle that was added to
+                // `inner.queued_entries` has valid authorization and satisfies the
+                // Sapling-specific consensus rules.
```

Comment:
It isn't documented (in https://github.com/zcash/librustzcash/blob/main/zcash_proofs/src/sapling/verifier/batch.rs) that `validate` must only be called if every `check_bundle` returned true. (This is not the same as saying that `validate` returns an unspecified result if any `check_bundle` returned false.) It should be documented.

Reply:
It's not a requirement of the `zcash_proofs` batch validator, but you're correct that it's now a requirement _here_ due to how the caching is added. What I meant by my comment is that given how we use this, we're fine; I'll document this on the method.